sufia-models/app/models/batch.rb

```diff
@@ -10,13 +10,29 @@ class Batch < ActiveFedora::Base
   property :status, predicate: ::RDF::DC.type
 
   def self.find_or_create(id)
-    # FIXME potential race condition in this method. Consider that `find' may raise
-    # ObjectNotFound in multiple processes. However, Fedora should raise an error
-    # if we try to create two objects with the same id.
     begin
       Batch.find(id)
     rescue ActiveFedora::ObjectNotFoundError
-      Batch.create(id: id)
+      safe_create(id)
     end
   end
+
+  private
+
+    # This method handles most race conditions gracefully. 
+    # If a batch with the same ID is created by another thread
+    # we fetch the batch that was created (rather than throwing
+    # an error) and continute.
+    def self.safe_create(id)
+      begin      
+        Batch.create(id: id)
+      rescue ActiveFedora::IllegalOperation => ex
+        if ex.message.starts_with?("Attempting to recreate existing ldp_source")
+          # This is OK. The batch was created in between the checks. 
+          Batch.find(id)
+        else
```

Comment:
Do you want this else only to occur when `ActiveFedora::IllegalOperation` is raised or when any other exception is raised?